### PR TITLE
Add DinoV2 for semantic segmentation

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by kprokofi** [comment]
Need to conduct more experiments and establish hyperparameters for DINOv2 config before setting this PR in review

**Comment by kprokofi** [comment]
I deleted BNHead since Linear Head performed on par or even a bit better in my experiments. Freezing backbone is helpful for fine tuning task + saves training time

Final results:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


  | Pascal-VOC (cut) | kvasir-small | kvasir large | cityscapes (cut)
-- | -- | -- | -- | --
DinoV2s | 81.33 | 83.77 | 86.54 | 61.82



</body>

</html>

File: src/otx/algo/segmentation/backbones/dinov2.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        self.backbone = torch.hub.load(repo_or_dir="facebookresearch/dinov2", model=name)
+        if freeze_backbone:
+            self._freeze_backbone(self.backbone)
+
+        # take intermediate layers to preserve spatial dimension
+        self.backbone.forward = partial(
+            self.backbone.get_intermediate_layers,
+            n=out_index,
+            reshape=True,
+     )
```
Comment (vinnamkim) [inline src/otx/algo/segmentation/backbones/dinov2.py]:
```suggestion
          )
```
@harimkang 
Again, there is no code formatter in the pre-commit hook.

File: src/otx/recipe/segmentation/dino_v2_seg.yaml [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+          type: SyncBN
+          requires_grad: true
+        align_corners: false
+        loss_decode:
+          type: CrossEntropyLoss
+          use_sigmoid: false
+          loss_weight: 1.0
+      pretrained: null
+      test_cfg:
+        mode: whole
```
Comment (vinnamkim) [inline src/otx/recipe/segmentation/dino_v2_seg.yaml:118]: I investigated the configuration in this notebook, https://github.com/facebookresearch/dinov2/blob/main/notebooks/semantic_segmentation.ipynb.
However, what's the reason for the difference in `test_cfg`?
image
You used `mode: whole` but the notebook used `mode: slide`.

File: src/otx/algo/segmentation/backbones/dinov2.py (hunk 1 above)
Comment (harimkang) [reply src/otx/algo/segmentation/backbones/dinov2.py]: I'll look into it and give you some suggestions, thanks for letting me know.

File: src/otx/recipe/segmentation/dino_v2_seg.yaml (hunk 2 above)
Comment (kprokofi) [reply src/otx/recipe/segmentation/dino_v2_seg.yaml:118]: I aligned testing configuration with other our segmentation models. There shouldn't be any differences in final metric. I also aligned the resolution to our standard resolution that we use in Geti.

Comment (vinnamkim) [reply src/otx/recipe/segmentation/dino_v2_seg.yaml:118]: image

https://mmsegmentation.readthedocs.io/en/latest/advanced_guides/models.html#prediction-modes

I can see here that `test_cfg` really affects to the model `forward()` in the prediction. It seems being able to affect the final metric.

For VisionTransformer, input size is more important than ConvNet. This is because the input size affects to the token sequence length and it affects to the both sides, model accuracy and performance:

1) Self attention computation is `O(n^2)`, `n` is token sequence length.
2) If the pretrained model had been training for `n` tokens, but you gave `m > n` tokens after that, the position embedding for `k=n+1,...,m` might not have been learnt well.

Comment (kprokofi) [reply src/otx/recipe/segmentation/dino_v2_seg.yaml:118]: Okay, let me check if there will be the difference

**Comment by kprokofi** [comment]
@vinnamkim 
I fixed seeds and set deterministic True for fair experiments. I conducted training with different settings to compare data resolution and inference methods (whole vs slide).
There is no difference in metric between the slide and the whole inference.
Also, there is almost no difference in metric for two resolutions (560 and 640). However, the model is slower with higher resolution as expected

I propose to align with our other templates and keep current settings for DINOV2

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/kprokofi/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


Model | VOC | Cityscapes | Kvasir-small | VOC, e2e time | Cityscapes, e2e time | Kvasir-small, e2e time | AVG mIoU | AVG Time
-- | -- | -- | -- | -- | -- | -- | -- | --
slide_640 | 81.21 | 62.85 | 83.29 | 2130 | 3215 | 100 | 75.78 | 1815.00
slide_560 | 81.66 | 61.32 | 84.33 | 1712 | 2980 | 87 | 75.77 | 1593.00
whole_640 | 81.21 | 62.85 | 83.29 | 2100 | 3184 | 99 | 75.78 | 1794.33
whole_560 | 81.66 | 61.32 | 84.33 | 1695 | 2947 | 87 | 75.77 | 1576.33



</body>

</html>

File: src/otx/recipe/segmentation/dino_v2_seg.yaml (hunk 2 above)
Comment (kprokofi) [reply src/otx/recipe/segmentation/dino_v2_seg.yaml:118]: So, considering my experiments, I stick to current settings

**Review by vinnamkim — DISMISSED** [review]

**Review by jaegukhyun — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]

**Comment by kprokofi** [comment]
Rebased to V2 branch to resolve conflicts

**Review by vinnamkim — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]
